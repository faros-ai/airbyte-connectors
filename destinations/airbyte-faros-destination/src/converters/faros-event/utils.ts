import {ErrorObject} from 'ajv';
import {groupBy} from 'lodash';

import {Artifact, FarosPath, FarosTag, RunStep} from './types';

interface OrganizationKey {
  readonly uid: string;
  readonly source: string;
}

export interface DeploymentKey {
  readonly uid: string;
  readonly source: string;
}

export interface DeployIds {
  readonly id: string;
  readonly source: string;
  readonly environment: string;
  readonly application: string;
}

export interface ArtifactKey {
  readonly uid: string;
  readonly repository: {
    readonly uid: string;
    readonly organization: OrganizationKey;
  };
}

export interface ArtifactIds {
  readonly id: string;
  readonly repository: string;
  readonly organization: string;
  readonly source: string;
}

export interface CommitKey {
  readonly sha: string;
  readonly repository: {
    readonly name: string;
    readonly organization: OrganizationKey;
  };
}

export interface CommitIds {
  readonly sha: string;
  readonly repository: string;
  readonly organization: string;
  readonly source: string;
}

export interface BuildKey {
  readonly uid: string;
  readonly pipeline: {
    readonly uid: string;
    readonly organization: OrganizationKey;
  };
}

export interface BuildIds {
  readonly id: string;
  readonly pipeline: string;
  readonly organization: string;
  readonly source: string;
}

export interface BuildStepKey {
  readonly uid: string;
  readonly build: BuildKey;
}

export interface AgentKey {
  readonly uid: string;
  readonly source: string;
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export function resolveCommitIds(commit: any): CommitIds {
  if (commit.uri) {
    const parsedUri = parseUri(commit.uri);
    return {
      sha: parsedUri[3],
      repository: parsedUri[2].toLowerCase(),
      organization: parsedUri[1].toLowerCase(),
      source: parsedUri[0],
    };
  }
  return {
    sha: commit.sha,
    repository: commit.repository.toLowerCase(),
    organization: commit.organization.toLowerCase(),
    source: commit.source,
  };
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export function makeCommitKey(commit: any): CommitKey {
  const commitIds = resolveCommitIds(commit);
  return {
    sha: commitIds.sha,
    repository: {
      name: commitIds.repository,
      organization: {
        uid: commitIds.organization,
        source: commitIds.source,
      },
    },
  };
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export function resolveBuildIds(run: any): BuildIds {
  if (run.uri) {
    const parsedUri = parseUri(run.uri);
    return {
      id: parsedUri[3],
      pipeline: parsedUri[2],
      organization: parsedUri[1],
      source: parsedUri[0],
    };
  }
  return {
    id: run.id,
    pipeline: run.pipeline,
    organization: run.organization,
    source: run.source,
  };
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export function makeBuildKey(run: any): BuildKey {
  const buildIds = resolveBuildIds(run);
  return {
    uid: buildIds.id,
    pipeline: {
      uid: buildIds.pipeline,
      organization: {
        uid: buildIds.organization,
        source: buildIds.source,
      },
    },
  };
}

export function makeBuildStepKey(step: RunStep, build: BuildKey): BuildStepKey {
  return {
    uid: step.id,
    build,
  };
}

export function resolveDeploymentIds(deploy: any): DeployIds {
  let deployIds;
  if (deploy.uri) {
    const parsedUri = parseUri(deploy.uri);
    deployIds = {
      id: parsedUri[3],
      environment: parsedUri[2],
      application: parsedUri[1],
      source: parsedUri[0],
    };
  } else {
    deployIds = {
      id: deploy.id,
      environment: deploy.environment,
      application: deploy.application,
      source: deploy.source,
    };
  }
  return deployIds;
}

function enumToMap(enumValues: string[]): Map<string, string> {
  return new Map(enumValues.map((value) => [value.toLowerCase(), value]));
}

function resolveEnum(
  enumValue: Map<string, string>,
  field: string,
  fieldDetail?: string | null
): {category: string; detail?: string | null} {
  const lowerField = field.toLowerCase();
  const value = enumValue.get(lowerField);
  if (value) {
    return {category: value, detail: fieldDetail};
  }
  return {
    category: 'Custom',
    detail: field.concat(fieldDetail ? ` - ${fieldDetail}` : ''),
  };
}

const deploymentEnvironments = enumToMap([
  'Prod',
  'Staging',
  'QA',
  'Dev',
  'Sandbox',
  'Canary',
  'Custom',
]);

export function resolveDeploymentEnvironment(
  environment: string,
  environmentDetails?: string | null
): {category: string; detail?: string | null} {
  return resolveEnum(deploymentEnvironments, environment, environmentDetails);
}

const deploymentStatuses = enumToMap([
  'Canceled',
  'Custom',
  'Failed',
  'Queued',
  'Running',
  'RolledBack',
  'Success',
]);

export function resolveDeploymentStatus(
  status?: string,
  statusDetails?: string | null
): {category: string; detail?: string | null} | undefined {
  if (!status) {
    return undefined;
  }
  return resolveEnum(deploymentStatuses, status, statusDetails);
}

const runStatuses = enumToMap([
  'Canceled',
  'Custom',
  'Failed',
  'Queued',
  'Running',
  'Success',
  'Unknown',
]);

export function resolveRunStatus(
  status?: string,
  statusDetails?: string | null
): {category: string; detail?: string | null} | undefined {
  if (!status) {
    return undefined;
  }
  return resolveEnum(runStatuses, status, statusDetails);
}

const runStepTypes = enumToMap(['Custom', 'Manual', 'Script']);

export function resolveRunStepType(
  type?: string,
  typeDetails?: string | null
): {category: string; detail?: string | null} | undefined {
  if (!type) {
    return undefined;
  }
  return resolveEnum(runStepTypes, type, typeDetails);
}

const computeAgentOS = enumToMap(['Linux', 'MacOS', 'Windows', 'Custom']);

export function resolveComputeAgentOS(
  os?: string,
  osDetails?: string | null
): {category: string; detail?: string | null} | undefined {
  if (!os) {
    return undefined;
  }
  return resolveEnum(computeAgentOS, os, osDetails);
}

const metricDefinitionTypes = enumToMap([
  'Numeric',
  'String',
  'Date',
  'Timestamp',
  'Duration',
  'Percentage',
  'Custom',
]);

export function resolveMetricDefinitionType(
  type?: string,
  typeDetails?: string | null
): {category: string; detail?: string | null} | undefined {
  if (!type) {
    return undefined;
  }
  return resolveEnum(metricDefinitionTypes, type, typeDetails);
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export function resolveArtifactIds(artifact: any): ArtifactIds {
  if (artifact.uri) {
    const parsedUri = parseUri(artifact.uri);
    return {
      id: parsedUri[3],
      repository: parsedUri[2].toLowerCase(),
      organization: parsedUri[1].toLowerCase(),
      source: parsedUri[0],
    };
  }
  return {
    id: artifact.id,
    repository: artifact.repository.toLowerCase(),
    organization: artifact.organization.toLowerCase(),
    source: artifact.source,
  };
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export function makeArtifactKey(artifact: any): ArtifactKey {
  const artifactIds = resolveArtifactIds(artifact);
  return {
    uid: artifactIds.id,
    repository: {
      uid: artifactIds.repository.toLowerCase(),
      organization: {
        uid: artifactIds.organization.toLowerCase(),
        source: artifactIds.source,
      },
    },
  };
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export function makeDummyArtifact(commit: any): Artifact {
  const commitKey = makeCommitKey(commit);
  return {
    id: commitKey.sha,
    repository: commitKey.repository.name,
    organization: commitKey.repository.organization.uid,
    source: commitKey.repository.organization.source,
  };
}

// Convert ISO-8601 string, number string, number, or 'Now' into milliseconds
export function getTime(date?: string | number): number | undefined {
  if (date) {
    if (typeof date === 'string') {
      if (date === 'Now') {
        return Date.now();
      }
      const dateInt = Number(date);
      if (!isNaN(dateInt)) {
        return new Date(dateInt).getTime();
      }
    }
    return new Date(date).getTime();
  }
  return undefined;
}

// Convert ISO-8601 string, number string, number, or 'Now' into ISO-8601 string
export function toDate(date?: string | number): string | undefined {
  if (date) {
    if (typeof date === 'string') {
      if (date === 'Now') {
        return new Date(Date.now()).toISOString();
      }
      const dateInt = Number(date);
      if (!isNaN(dateInt)) {
        return new Date(dateInt).toISOString();
      }
    }
    return new Date(date).toISOString();
  }
  return undefined;
}

export function parseTags(tags?: FarosTag[] | string): FarosTag[] | undefined {
  const parser = (input: string): FarosTag[] => {
    const resTags: FarosTag[] = [];
    const keyValues = input.split(',');
    for (const keyValue of keyValues) {
      const splitTag = keyValue.split(':');
      resTags.push({key: splitTag[0], value: splitTag[1]});
    }
    return resTags;
  };
  return parseArray(parser, tags);
}

export function parsePaths(
  paths?: FarosPath[] | string
): FarosPath[] | undefined {
  const parser = (input: string): FarosPath[] => {
    return input.split(',').map<FarosPath>((path) => {
      return {path};
    });
  };
  return parseArray(parser, paths);
}

export function parsePullRequestNumber(
  number: string | number
): number | undefined {
  const prNum = typeof number === 'string' ? Number(number) : number;
  return isNaN(prNum) ? undefined : prNum;
}

export function parseUri(uri: string): string[] {
  const components: string[] = [];
  const split1 = uri.split('://');
  components.push(split1[0]);
  split1[1].split('/').forEach((value) => components.push(value));
  return components;
}

export function parseStringArray(
  array?: string[] | string,
  separator = ','
): string[] | undefined {
  return parseArray((s) => s.split(separator), array);
}

function parseArray<T>(
  parser: (input: string) => T[],
  input?: T[] | string
): T[] | undefined {
  if (typeof input === 'string') {
    return parser(input);
  }
  return input;
}

export interface EventValidationError {
  readonly message: string;
  readonly details?: ReadonlyArray<string>;
  readonly allowedValues?: ReadonlyArray<string>;
  readonly additionalProperties?: ReadonlyArray<string>;
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export function removeUndefinedProperties(object: any): any {
  const result = {...object};
  Object.keys(result).forEach(
    (key) => result[key] === undefined && delete result[key]
  );
  return result;
}

export function formatErrors(
  errors?: ErrorObject[] | null,
  rootObject?: string
): ReadonlyArray<EventValidationError> {
  if (!errors) {
    return [];
  }

  const formattedErrors: EventValidationError[] = [];
  const groupedErrors = groupBy<ErrorObject>(errors, (err) => {
    return formatErrorDetailsMessage(err, rootObject);
  });

  for (const [message, errs] of Object.entries(groupedErrors)) {
    const details = errs.flatMap(
      (err) =>
        err.params?.errors?.map((v: any) =>
          formatErrorDetailsMessage(v, rootObject)
        ) ?? []
    );
    const additionalProperties = errs.flatMap(
      (err) => err.params?.additionalProperty ?? []
    );
    formattedErrors.push({
      message,
      details: details.length > 0 ? details : undefined,
      allowedValues: getAllowedValues(errs),
      additionalProperties:
        additionalProperties.length > 0 ? additionalProperties : undefined,
    });
  }
  return formattedErrors;
}

function formatErrorDetailsMessage(
  error: ErrorObject,
  rootObject = 'data'
): string {
  const prefix = error?.instancePath
    ? `${rootObject}${error.instancePath.replace(/\//g, '.')} `
    : `${rootObject} `;
  return `${prefix}${error.message}`;
}

function getAllowedValues(errors: ErrorObject[]): string[] | undefined {
  let result;
  if (errors?.length > 0) {
    result = errors[0].params?.allowedValues;
    if (!result && errors[0].params?.errors?.length > 0) {
      result = errors[0].params?.errors[0].params?.allowedValues;
    }
  }
  return result;
}

// Max length for free-form description text fields such as issue body
const MAX_TEXT_LENGTH = 1000;
export function cleanAndTruncate(
  str?: string,
  maxLength?: number
): string | null | undefined {
  if (!str) {
    return str;
  }
  const length = maxLength ?? MAX_TEXT_LENGTH;
  let result: string;
  if (str.length <= length) {
    result = str;
  } else {
    // If the last character is part of a unicode surrogate pair, include the next character
    const lastChar = str.codePointAt(length - 1) ?? 0;
    result =
      lastChar > 65535
        ? str.substring(0, length + 1)
        : str.substring(0, length);
  }
  // eslint-disable-next-line no-control-regex
  return result.replace(/\u0000/g, '');
}
