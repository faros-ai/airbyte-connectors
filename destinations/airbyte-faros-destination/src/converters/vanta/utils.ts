import fs from 'fs-extra';
import path from 'path';

// Function to find a specific parent directory
function findParentDirectory(
  currentDir: string,
  targetDirName: string
): string | null {
  let currentPath = currentDir;
  while (currentPath !== '/') {
    if (path.basename(currentPath) === targetDirName) {
      return currentPath;
    }
    currentPath = path.dirname(currentPath);
  }
  return null;
}

export function getQueryFromName(name: string): string {
  const crt_dir = __dirname;
  // We go back to the part of dir which ends with name:
  const targetDirPath = findParentDirectory(
    crt_dir,
    'airbyte-faros-destination'
  );
  if (!targetDirPath) {
    throw new Error('airbyte-faros-destination directory not found');
  }

  const gql_dir = path.join(targetDirPath, 'resources');

  const fn = `${name}.gql`;
  // Check if fn is in gql_dir:
  if (!fs.existsSync(`${gql_dir}/${fn}`)) {
    throw new Error(`Query ${fn} not found in ${gql_dir}`);
  }
  // We access the directory resources/gql_queries to get the query
  return fs.readFileSync(`${gql_dir}/${fn}`, 'utf8');
}

export function looksLikeGitCommitSha(sha: string): boolean {
  return /^[a-f0-9]{40}$/i.test(sha);
}
