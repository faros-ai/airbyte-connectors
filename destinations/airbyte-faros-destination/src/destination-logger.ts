import pQueue from '@esm2cjs/p-queue';
import crypto from 'crypto';
import {
  AirbyteLog,
  AirbyteLogger,
  AirbyteLogLevel,
  AirbyteLogLevelOrder,
  AirbyteMessage,
  AirbyteSourceLog,
  isAirbyteLog,
  shouldWriteLog,
} from 'faros-airbyte-cdk';
import fs from 'fs';
import os from 'os';
import readline from 'readline';

export class LogFiles {
  private readonly srcPath: string;
  private readonly dstPath: string;

  private readonly srcStream: fs.WriteStream;
  private readonly dstStream: fs.WriteStream;

  private readonly srcWriteQueue: pQueue = new pQueue({concurrency: 1});
  private readonly dstWriteQueue: pQueue = new pQueue({concurrency: 1});

  constructor(private readonly logger: FarosDestinationLogger) {
    const logDir = os.tmpdir();
    this.srcPath = `${logDir}/src.log`;
    this.dstPath = `${logDir}/dst.log`;
    this.srcStream = fs.createWriteStream(this.srcPath, {
      flags: 'a',
      encoding: 'utf-8',
    });
    this.dstStream = fs.createWriteStream(this.dstPath, {
      flags: 'a',
      encoding: 'utf-8',
    });
  }

  writeSourceLogs(...logs: AirbyteSourceLog[]): void {
    this.srcWriteQueue.add(() => this.writeLogs(this.srcStream, logs));
  }

  writeDestinationLogs(...logs: AirbyteSourceLog[]): void {
    this.dstWriteQueue.add(() => this.writeLogs(this.dstStream, logs));
  }

  private logError(msg: string, error: any): void {
    this.logger.localWrite(
      AirbyteLog.make(AirbyteLogLevel.WARN, `${msg}: ${JSON.stringify(error)}`)
    );
  }

  private writeLogs(
    stream: fs.WriteStream,
    logs: AirbyteSourceLog[]
  ): Promise<void> | void {
    try {
      if (stream.closed || !logs.length) {
        return;
      }
      const logString =
        logs.map((log) => JSON.stringify(log)).join('\n') + '\n';
      if (!stream.write(logString)) {
        // If the stream returns false, wait for the drain event before continuing
        return new Promise((resolve) => {
          // Update maxListeners to suppress nodejs warning
          stream.setMaxListeners(stream.getMaxListeners() + 1);
          stream.once('drain', () => {
            stream.setMaxListeners(Math.max(stream.getMaxListeners() - 1, 0));
            resolve();
          });
        });
      }
    } catch (error) {
      this.logError('Failed to save logs to disk', error);
    }
  }

  async sortedLogs(
    logger?: AirbyteLogger
  ): Promise<{content: string; hash: string} | undefined> {
    try {
      await Promise.all([
        this.srcWriteQueue.onIdle(),
        this.dstWriteQueue.onIdle(),
      ]);

      this.srcStream.end();
      this.dstStream.end();

      await Promise.all([
        new Promise((resolve) => this.srcStream.once('close', resolve)),
        new Promise((resolve) => this.dstStream.once('close', resolve)),
      ]);

      const srcLogs: AirbyteSourceLog[] = [];
      const dstLogs: AirbyteSourceLog[] = [];

      logger?.debug('Gathering sync logs for uploading to Faros');
      await Promise.all([
        readFile(this.srcPath, (line) =>
          processLogLineFromFile(line, srcLogs, logger)
        ),
        readFile(this.dstPath, (line) =>
          processLogLineFromFile(line, dstLogs, logger)
        ),
      ]);
      const allLogs = srcLogs.concat(dstLogs);
      allLogs.sort((a, b) => a.timestamp - b.timestamp);
      const content = allLogs.map((log) => JSON.stringify(log)).join('\n');
      const hash = crypto.createHash('md5').update(content).digest('base64');
      logger?.debug('Finished gathering sync logs');
      return {content, hash};
    } catch (error) {
      this.logError('Failed to gather logs for uploading to Faros', error);
      return undefined;
    }
  }
}

function processLogLineFromFile(
  line: string,
  logArray: AirbyteSourceLog[],
  logger?: AirbyteLogger
): AirbyteSourceLog {
  try {
    const log = JSON.parse(line);
    logArray.push(log);
  } catch (err: any) {
    logger?.warn(
      `Failed to parse log line: ${line}. Error: ${JSON.stringify(err)}`
    );
    return undefined;
  }
}

async function readFile(
  filePath: string,
  processLine: (line: string) => void
): Promise<void> {
  return new Promise<void>((resolve, reject) => {
    const fileStream = fs.createReadStream(filePath, 'utf-8');

    const rl = readline.createInterface({
      input: fileStream,
      crlfDelay: Infinity,
    });

    rl.on('line', (line) => processLine(line));
    rl.on('close', () => resolve());
    fileStream.on('error', (error) => reject(error));
  });
}

export class FarosDestinationLogger extends AirbyteLogger {
  private pendingLogs: AirbyteSourceLog[] = [];
  private _logFiles: LogFiles;

  // Default to true to ensure initial logs are saved until destination can
  // determine if it should upload logs to Faros
  private _shouldSaveLogs: boolean = true;

  constructor(level?: AirbyteLogLevel) {
    super(level);
  }

  set logFiles(logFiles: LogFiles) {
    this._logFiles = logFiles;
    if (this._logFiles && this._shouldSaveLogs) {
      this._logFiles.writeDestinationLogs(...this.pendingLogs);
      this.pendingLogs.length = 0;
    }
  }

  get shouldSaveLogs(): boolean {
    return this._shouldSaveLogs;
  }

  set shouldSaveLogs(shouldSaveLogs: boolean) {
    this._shouldSaveLogs = shouldSaveLogs;
    if (!shouldSaveLogs) {
      this.pendingLogs.length = 0;
    }
  }

  localWrite(msg: AirbyteMessage): void {
    super.write(msg);
  }

  override write(msg: AirbyteMessage): void {
    super.write(msg);

    if (
      this._shouldSaveLogs &&
      isAirbyteLog(msg) &&
      shouldWriteLog(msg, this.level)
    ) {
      const syncLog: AirbyteSourceLog = {
        timestamp: Date.now(),
        message: {
          level: AirbyteLogLevelOrder(msg.log.level),
          msg: msg.log.message,
        },
      };

      if (this._logFiles) {
        this._logFiles.writeDestinationLogs(syncLog);
      } else {
        this.pendingLogs.push(syncLog);
      }
    }
  }
}
