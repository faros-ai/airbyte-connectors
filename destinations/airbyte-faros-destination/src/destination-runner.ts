import {Command} from 'commander';
import {
  AirbyteConfig,
  AirbyteDestinationRunner,
  AirbyteLogger,
  AirbyteSpec,
} from 'faros-airbyte-cdk';

import {DestinationConfig} from './common/types';
import {Converter} from './converters/converter';
import {ConverterRegistry} from './converters/converter-registry';
import {FarosDestination} from './destination';

/** Faros Destination Runner */
export class FarosDestinationRunner extends AirbyteDestinationRunner<DestinationConfig> {
  readonly program: Command;

  /**
   * Creates a new Faros Destination Runner instance
   *
   * @param specOverride overrides the default specification with a custom one
   */
  constructor(specOverride: AirbyteSpec = undefined) {
    const logger = new AirbyteLogger();
    const destination = new FarosDestination(logger, specOverride);
    super(logger, destination);
    this.program = super.mainCommand();
  }

  /**
   * Register converters for custom streams
   */
  registerConverters(...converters: ReadonlyArray<Converter>): void {
    converters.forEach(ConverterRegistry.addConverter);
  }

  /**
   * Add an additional configuration check. Handy when using spec override.
   */
  onConfigCheck(check: (config: AirbyteConfig) => Promise<void>): void {
    (this.destination as FarosDestination).onConfigCheck = check;
  }
}
