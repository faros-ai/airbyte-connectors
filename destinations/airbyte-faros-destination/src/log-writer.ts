import crypto from 'crypto';
import {AirbyteSourceLog} from 'faros-airbyte-cdk';
import fs from 'fs';
import os from 'os';
import readline from 'readline';

export class LogWriter {
  private readonly sourceLogPath: string;
  private readonly destLogPath: string;

  private readonly sourceLogStream: fs.WriteStream;
  private readonly destLogStream: fs.WriteStream;

  constructor() {
    const logDir = os.tmpdir();
    this.sourceLogPath = `${logDir}/source.log`;
    this.destLogPath = `${logDir}/dest.log`;
    this.sourceLogStream = fs.createWriteStream(this.sourceLogPath, {
      flags: 'a',
    });
    this.destLogStream = fs.createWriteStream(this.destLogPath, {flags: 'a'});
  }

  async writeSourceLog(sourceLog: AirbyteSourceLog): Promise<void> {
    await writeDataHandlingBackpressure(
      this.sourceLogStream,
      JSON.stringify(sourceLog) + '\n'
    );
  }

  async writeDestLog(sourceLog: AirbyteSourceLog): Promise<void> {
    await writeDataHandlingBackpressure(
      this.destLogStream,
      JSON.stringify(sourceLog) + '\n'
    );
  }

  async sortedLogs(): Promise<{content: string; hash: string}> {
    this.sourceLogStream.end();
    this.destLogStream.end();

    const sourceLogs: AirbyteSourceLog[] = [];
    const destLogs: AirbyteSourceLog[] = [];

    await Promise.all([
      readFileLineByLine(this.sourceLogPath, (line) => {
        sourceLogs.push(JSON.parse(line));
      }),
      readFileLineByLine(this.destLogPath, (line) => {
        destLogs.push(JSON.parse(line));
      }),
    ]);

    const sortedLogs = sourceLogs.concat(destLogs).sort((a, b) => {
      return a.timestamp - b.timestamp;
    });
    const content = sortedLogs.map((log) => JSON.stringify(log)).join('\n');
    const hash = crypto.createHash('md5').update(content).digest('base64');
    return {content, hash};
  }
}

async function readFileLineByLine(
  filePath: string,
  processLine: (line: string) => void
): Promise<void> {
  return new Promise<void>((resolve, reject) => {
    const fileStream = fs.createReadStream(filePath);

    const rl = readline.createInterface({
      input: fileStream,
      crlfDelay: Infinity,
    });

    rl.on('line', (line) => {
      processLine(line);
    });

    rl.on('close', () => {
      console.log('Finished reading the file.');
      resolve(); // Resolve the promise when file reading is complete
    });

    fileStream.on('error', (error) => {
      console.error(
        `An error occurred while reading the file: ${error.message}`
      );
      reject(error); // Reject the promise if an error occurs
    });
  });
}

async function writeDataHandlingBackpressure(
  stream: fs.WriteStream,
  data: any
): Promise<void> {
  if (!stream.write(data)) {
    // If the stream returns false, wait for the drain event before continuing
    await waitForDrain(stream);
  }
}

function waitForDrain(stream: fs.WriteStream): Promise<void> {
  return new Promise((resolve) => {
    stream.once('drain', resolve);
  });
}
