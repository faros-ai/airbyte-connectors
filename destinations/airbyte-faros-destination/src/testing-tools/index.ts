export * from './cli';
export * from './testing-tools';
export * from './utils';
