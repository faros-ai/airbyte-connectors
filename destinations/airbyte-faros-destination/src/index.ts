import {Command} from 'commander';

import {FarosDestinationRunner} from './destination-runner';

/**
 * The main entry point
 *
 * You can optionally provide options `exitOverride` and `configureOutput` to
 * override default behavior of the CLI when it detects an error and where it
 * writes to respectively. See https://github.com/tj/commander.js#override-exit-and-output-handling
 * for detailed info on how these options work.
 */
export function mainCommand(options?: {
  exitOverride?: boolean;
  suppressOutput?: boolean;
}): Command {
  const destinationRunner = new FarosDestinationRunner();
  const program = destinationRunner.mainCommand();

  if (options?.exitOverride) {
    program.exitOverride();
  }
  if (options?.suppressOutput) {
    program.configureOutput({
      // eslint-disable-next-line @typescript-eslint/no-empty-function
      writeOut: () => {},
      // eslint-disable-next-line @typescript-eslint/no-empty-function
      writeErr: () => {},
    });
  }

  return program;
}

export {Edition, InvalidRecordStrategy} from './common/types';
export {
  Converter,
  DestinationRecord,
  DestinationModel,
  StreamContext,
  StreamName,
  StreamNameSeparator,
} from './converters/converter';
export {FarosDestinationRunner} from './destination-runner';
