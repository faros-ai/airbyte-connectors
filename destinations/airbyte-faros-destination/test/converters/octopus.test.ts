import _ from 'lodash';
import {getLocal} from 'mockttp';

import {Edition, InvalidRecordStrategy} from '../../src';
import {CLI, read} from '../cli';
import {initMockttp, tempConfig, testLogger} from '../testing-tools';
import {octopusAllStreamsLog} from './data';
import {assertProcessedAndWrittenModels} from "./utils";

describe('octopus', () => {
  const logger = testLogger();
  const mockttp = getLocal({debug: false, recordTraffic: false});
  const catalogPath = 'test/resources/octopus/catalog.json';
  let configPath: string;
  const streamNamePrefix = 'mytestsource__octopus__';

  beforeEach(async () => {
    await initMockttp(mockttp);
    configPath = await tempConfig(
      mockttp.url,
      InvalidRecordStrategy.SKIP,
      Edition.CLOUD,
      {},
      {}
    );
  });

  afterEach(async () => {
    await mockttp.stop();
  });

  test('process records from all streams', async () => {
    const cli = await CLI.runWith([
      'write',
      '--config',
      configPath,
      '--catalog',
      catalogPath,
      '--dry-run',
    ]);
    cli.stdin.end(octopusAllStreamsLog, 'utf8');

    const stdout = await read(cli.stdout);
    logger.debug(stdout);

    const processedByStream = {
      deployments: 6,
      releases: 6,
    };
    const processed = _(processedByStream)
      .toPairs()
      .map((v) => [`${streamNamePrefix}${v[0]}`, v[1]])
      .orderBy(0, 'asc')
      .fromPairs()
      .value();

    const writtenByModel = {
      cicd_Deployment: 6,
      cicd_Release: 6,
    };

    await assertProcessedAndWrittenModels(processedByStream, writtenByModel, stdout, processed, cli);
  });
});
