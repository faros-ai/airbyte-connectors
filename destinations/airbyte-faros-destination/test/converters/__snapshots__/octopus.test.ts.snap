// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`octopus basic test process records from all streams 1`] = `
Array [
  "Processed 12 records",
  "Processed records by stream: {\\\\\\"mytestsource__octopus__deployments\\\\\\":6,\\\\\\"mytestsource__octopus__releases\\\\\\":6}\\"},\\"type\\":\\"LOG\\"}",
  "Would write 12 records",
  "Would write records by model: {\\\\\\"cicd_Deployment\\\\\\":6,\\\\\\"cicd_Release\\\\\\":6}\\"},\\"type\\":\\"LOG\\"}",
  "Skipped 0 records",
  "Errored 0 records",
]
`;
